[Blog] OpenSearch MCP Server: Empowering EC2 Team's Droplet Analysis (#3909)
opensearch-project/project-website: _posts/2025-08-18-opensearch-mcp-server-empowering-ec2-teams-droplet-analysis.md

Next, they created a YAML configuration file (clusters.yml) defining all their OpenSearch domains:

```yaml
version: "1.0"
description: "OpenSearch cluster configurations"

clusters:
    # Cluster for hardware engineering data
    hardware-logs:
    opensearch_url: "https://<Hardware cluster URL>"
    profile: "hardware-profile"
    aws_region: "us-east-1"
    
    # Cluster for droplet metrics
    droplet-metrics:
    opensearch_url: "https://<Metrics cluster URL>"
    profile: "metrics-profile"
    aws_region: "us-west-2"
    
    # Cluster for system events
    system-events:
    opensearch_url: "https://<Events cluster URL>"
    profile: "events-profile"
    aws_region: "eu-west-1"
```

The authentication for each cluster was managed through AWS profiles, allowing the OpenSearch MCP server to use the appropriate credentials for each cluster when needed.

### AWS Credentials File Structure

To complement the cluster configuration, here's how the AWS credentials file (~/.aws/credentials) was structured to support the profile-based authentication:

```ini
[hardware-profile]
aws_access_key_id = <your-access-key-id>
aws_secret_access_key = <your-secret-access-key>

> aws_session_token = <your-session-token>

Reviewer: Wouldn't session token expire after few hours? is this how EC2 team implemented their solution?

Author: Yes, they have a secure internal script which periodically reloads the credentials.

region = us-east-1

[metrics-profile]
aws_access_key_id = <your-access-key-id>
aws_secret_access_key = <your-secret-access-key>
aws_session_token = <your-session-token>
region = us-west-2

[events-profile]
aws_access_key_id = <your-access-key-id>
aws_secret_access_key = <your-secret-access-key>
aws_session_token = <your-session-token>
region = us-west-2
```

**This profile-based approach allowed the EC2 team to:**
* Leverage their existing AWS IAM roles and permissions
* Maintain credential separation between different environments
* Simplify credential rotation without modifying configuration files
* Ensure proper access controls for each OpenSearch domain

The OpenSearch MCP server supports various other authentication methods like basic authentication (username/password), direct IAM role authentication etc. Explore more authentication options in the [OpenSearch MCP Server documentation](https://github.com/opensearch-project/opensearch-mcp-server-py/blob/main/USER_GUIDE.md#authentication)

> ## Step 3: Provide Context to Amazon Q

Reviewer: Should we also call out that Amazon Q could be replaced with any other agentic framework like langchain?

Author: Makes sense will add that

date: 2025-08-18
categories:
    - technical-posts
meta_keywords: "OpenSearch MCP server, Amazon Q, EC2 droplet analysis, multi-cluster connectivity, operational automation, Model Context Protocol"
meta_description: "Learn how Amazon's EC2 team implemented OpenSearch MCP server with Amazon Q to automate droplet failure analysis across multiple OpenSearch domains, reducing investigation time from hours to minutes."
twittercard:
    description: "Amazon EC2 team revolutionized droplet quality investigations using OpenSearch MCP server with Amazon Q, transforming manual processes into efficient automated workflows."
    image: /assets/img/opensearch-logo-themed.svg
    image_alt: "OpenSearch Logo"
---

# Introduction

Amazon EC2 team recently leveraged OpenSearch's Model Context Protocol (MCP) server to revolutionize their droplet quality investigations. By connecting Amazon Q with multiple OpenSearch domains simultaneously, they've transformed a previously manual, time-consuming process into an efficient, automated workflow. This blog explores how the EC2 team implemented this solution, the technical setup involved, and the impressive results they've achieved.

# The Challenge

> The EC2 team manages droplet quality across their fleet, investigating failures and impairments to maintain reliability. Their data was distributed across multiple OpenSearch domains in various regions, each storing logs for different components. Engineers faced a multi-step process when troubleshooting:

Reviewer 2: "OpenSearch domain" generally used in AWS OpenSearch context. Should we use a more general word "OpenSearch cluster"?

* Search in one domain to find initial information
* Use those findings to query another domain
* Correlate data across multiple sources
* Repeat until root cause is identified

> This process required extensive domain knowledge about where to search and what patterns to look for. The team recognized that their engineers' expertise could be better utilized for solving issues rather than manually gathering and correlating data.

Reviewer 2: ```suggestion
This manual  process required deep knowledge of data locations and search patterns. The team realized their engineers' time would be better spent solving problems rather than hunting for and correlating data across systems.
```

* Search in one cluster to find initial information
* Use those findings to query another cluster
* Correlate data across multiple sources
* Repeat until root cause is identified

This manual process required deep knowledge of data locations and search patterns. The team realized their engineers' time would be better spent solving problems rather than hunting for and correlating data across systems.

# The Solution

The EC2 team implemented a solution combining:
* **Amazon Q CLI** as the intelligent agent to analyze data and provide insights

> **Note:** While this blog demonstrates the solution with Amazon Q CLI, you can replace it with any other agentic solution such as Claude, OpenAI, or other AI agents that support the Model Context Protocol
* **OpenSearch MCP Server** as the connectivity layer to multiple OpenSearch clusters
* **EC2 instance** running the agent with an attached instance role for secure cluster access

> * **IAM-based Authentication** for secure authentication to different clusters

Reviewer 3: ```suggestion
* **AWS Identity and Access Management (IAM)-based authentication** for secure authentication to different clusters.
```

> ![OpenSearch MCP Server Architecture](/assets/media/blog-images/2025-08-18-opensearch-mcp-server-empowering-ec2-teams-droplet-analysis/mcp-architecture-diagram.png){: .img-fluid}

Reviewer 3: The image should be introduced/referenced in text preceding it. What does the image show the reader?

Author: The image explains how the components mentioned before interact with each other. Will update to this:

```
The following architecture diagram illustrates how these components work together:

![OpenSearch MCP Server Architecture](/assets/media/blog-images/2025-08-18-opensearch-mcp-server-empowering-ec2-teams-droplet-analysis/mcp-architecture-diagram.png){: .img-fluid}

*The architecture shows Amazon Q CLI connecting to an EC2 instance that runs the OpenSearch MCP server. The OpenSearch MCP server retrieves credentials from the EC2 instance profile and dynamically assumes the appropriate IAM role at runtime based on the query, providing access to the corresponding OpenSearch cluster (hardware-logs, droplet-metrics, or system-events).*
```

> The key innovation was leveraging OpenSearch MCP Server's multi-cluster connectivity feature, which allowed them to interact with multiple OpenSearch clusters through a single interface, avoiding tool proliferation that could overwhelm the agent.

Reviewer 3: ```suggestion
The key innovation was leveraging the OpenSearch MCP server's multi-cluster connectivity feature, which allowed them to interact with multiple OpenSearch clusters through a single interface, avoiding tool proliferation that could overwhelm the agent.
```

> 

Reviewer 3: Line 47: Does "them" refer to "the EC2 team"? If yes, please replace.

> # Implementation Steps

Reviewer 3: ```suggestion
## Implementation steps
```

> ## Step 1: Set Up IAM Authentication

Reviewer 3: ```suggestion
### Step 1: Set up IAM authentication
```

The EC2 team's infrastructure spans three OpenSearch clusters across different AWS accounts:

1. **hardware-logs**: Stores hardware engineering data and failure events
2. **droplet-metrics**: Contains performance metrics and system telemetry  

> 3. **system-events**: Holds lifecycle events and administrative actions

Reviewer 3: ```suggestion
3. **system-events**: Stores lifecycle events and administrative actions
```

> To securely access these distributed clusters, the team implemented an EC2 instance-based authentication solution that allows them to connect to all three clusters from a single location while maintaining proper security controls.

Reviewer 3: ```suggestion
To securely access these distributed clusters, the team implemented an EC2-instance-based authentication solution that allows them to connect to all three clusters from a single location while maintaining proper security controls.
```

> ### IAM Setup Steps:

Reviewer 3: ```suggestion
**IAM setup steps**:
```

> #### 1.1 Create the Three Cluster Access Roles (in Cluster Accounts)

Reviewer 3: ```suggestion
1. Create the three cluster access roles (in cluster accounts): First, the EC2 team created the three IAM roles, in their respective AWS accounts, that would have access to each OpenSearch cluster. Each role needed a trust policy that would allow the EC2 instance role from the main account to assume it.
```

> First, the EC2 team created the three IAM roles in their respective AWS accounts that would have access to each OpenSearch cluster. Each role needed a trust policy that would allow the EC2 instance role from the main account to assume it.

Reviewer 3: ```suggestion
```

> **Example: Hardware Logs Cluster Role (Account: 123456789012):**

Reviewer 3: ```suggestion
**Example -- Hardware logs cluster role (Account: 123456789012)**:
```

```json
{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Effect": "Allow",
      "Principal": {
        "AWS": "arn:aws:iam::111111111111:role/EC2InstanceRole"
      },
      "Action": "sts:AssumeRole"
    }
  ]
}
```

> **Note:** The team created similar trust policies for the remaining two cluster roles in their respective accounts (234567890123 and 345678901234).

Reviewer 3: ```suggestion
**Note**: The team created similar trust policies for the remaining two cluster roles in their respective accounts (234567890123 and 345678901234).
```

Each role also needed a permissions policy for OpenSearch access:

```json
{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Effect": "Allow",
      "Action": [
        "es:ESHttpGet",
        "es:ESHttpPost",
        "es:ESHttpPut",
        "es:ESHttpDelete"
      ],
      "Resource": "arn:aws:es:REGION:ACCOUNT_ID:domain/CLUSTER_NAME/*"
    }
  ]
}
```

> #### 1.2 Create the EC2 Instance Role

Reviewer 3: ```suggestion
1. Create the EC2 instance role.
```

The team created an IAM role for the EC2 instance with the following trust policy:

```json
{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Effect": "Allow",
      "Principal": {
        "Service": "ec2.amazonaws.com"
      },
      "Action": "sts:AssumeRole"
    }
  ]
}
```

> #### 1.3 Attach Assume Role Policy

Reviewer 3: ```suggestion
1. Attach an assume role policy.

```

The team attached the following policy to the EC2 instance role to allow it to assume the cluster access roles across different AWS accounts:

```json
{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Effect": "Allow",
      "Action": "sts:AssumeRole",
      "Resource": [
        "arn:aws:iam::123456789012:role/hardwareLogsClusterAccessRole",
        "arn:aws:iam::234567890123:role/dropletMetricsClusterAccessRole",
        "arn:aws:iam::345678901234:role/systemEventsClusterAccessRole"
      ]
    }
  ]
}
```

> **Note:** This policy allows the EC2 instance role to assume roles in the three cluster accounts.

Reviewer 3: ```suggestion
**Note**: This policy allows the EC2 instance role to assume roles in the three cluster accounts.
```

> #### 1.4 Create Instance Profile

Reviewer 3: ```suggestion
1. Create an instance profile: The EC2 team created an instance profile and attached it to their EC2 instance during launch.
```

  system-events:
    opensearch_url: "https://<Events cluster URL>"
    iam_arn: "arn:aws:iam::345678901234:role/systemEventsClusterAccessRole" 
    aws_region: "eu-west-1"
```

The authentication for each cluster was managed through the instance role attached to the EC2 instance. Since the clusters were in different AWS accounts, the OpenSearch MCP Server used the following authentication flow:

* Gets credentials from the attached IAM instance role
* Assumes the provided IAM ARN for each cluster across account boundaries

**Note:** This authentication flow works because the following trust relationships were established during infrastructure setup:
* EC2 instance role needs permissions to assume target IAM roles (configured in Step 1.3)
* Each target role needs to trust the EC2 instance role (configured in Step 1.1)

**Key Benefits of This Solution:**
* **Enhanced Security**: No static credentials stored on instances
* **Simplified Management**: Centralized access control through IAM roles
* **AWS Best Practices**: Follows recommended security patterns for production environments
* **Automatic Credential Rotation**: Credentials are refreshed automatically by AWS

> The OpenSearch MCP Server supports various other authentication methods like basic authentication (username/password), direct AWS credentials authentication, etc. Explore more authentication options in the [OpenSearch MCP Server documentation](https://github.com/opensearch-project/opensearch-mcp-server-py/blob/main/USER_GUIDE.md#authentication)

Reviewer 3: Ok, my change just reflects the actual title of the document, which includes "Python".

* Search in one cluster to find initial information.
* Use those findings to query another cluster.
* Correlate data across multiple sources.
* Repeat until the root cause is identified.

This manual process required deep knowledge of data locations and search patterns. The team realized their engineers' time would be better spent solving problems rather than hunting for and correlating data across systems.

## The solution

The EC2 team implemented a solution combining:
* **Amazon Q CLI** as the intelligent agent for analyzing data and providing insights.

> **Note**: While this post demonstrates the solution with Amazon Q CLI, you can also use any other agentic solution, such as Claude, OpenAI, or other AI agents that support MCP.
* **OpenSearch MCP server** as the connectivity layer for multiple OpenSearch clusters.
* An **EC2 instance** running the agent with an attached instance role for secure cluster access.
* **AWS Identity and Access Management (IAM)-based authentication** for secure authentication to different clusters.

> The following architecture diagram illustrates how these components work together:

Reviewer 3: ```suggestion
The following architecture diagram illustrates how these components work together.
```

**IAM setup steps**:

1. Create the three cluster access roles (in cluster accounts): First, the EC2 team created the three IAM roles, in their respective AWS accounts, that would have access to each OpenSearch cluster. Each role needed a trust policy that would allow the EC2 instance role from the main account to assume it.

**Example -- Hardware logs cluster role (Account: 123456789012)**:
```json
{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Effect": "Allow",
      "Principal": {
        "AWS": "arn:aws:iam::111111111111:role/EC2InstanceRole"
      },
      "Action": "sts:AssumeRole"
    }
  ]
}
```

**Note**: The team created similar trust policies for the remaining two cluster roles in their respective accounts (234567890123 and 345678901234).

Each role also needed a permissions policy for OpenSearch access:

```json
{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Effect": "Allow",
      "Action": [
        "es:ESHttpGet",
        "es:ESHttpPost",
        "es:ESHttpPut",
        "es:ESHttpDelete"
      ],
      "Resource": "arn:aws:es:REGION:ACCOUNT_ID:domain/CLUSTER_NAME/*"
    }
  ]
}
```

> 1. Create the EC2 instance role.

Reviewer 3: ```suggestion
2. Create the EC2 instance role.
```

The team created an IAM role for the EC2 instance with the following trust policy:

```json
{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Effect": "Allow",
      "Principal": {
        "Service": "ec2.amazonaws.com"
      },
      "Action": "sts:AssumeRole"
    }
  ]
}
```

1. Attach an assume role policy.

The team attached the following policy to the EC2 instance role to allow it to assume the cluster access roles across different AWS accounts:

```json
{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Effect": "Allow",
      "Action": "sts:AssumeRole",
      "Resource": [
        "arn:aws:iam::123456789012:role/hardwareLogsClusterAccessRole",
        "arn:aws:iam::234567890123:role/dropletMetricsClusterAccessRole",
        "arn:aws:iam::345678901234:role/systemEventsClusterAccessRole"
      ]
    }
  ]
}
```

**Note**: This policy allows the EC2 instance role to assume roles in the three cluster accounts.

> 1. Create an instance profile: The EC2 team created an instance profile and attached it to their EC2 instance during launch.

Reviewer 3: ```suggestion
4. Create an instance profile: The EC2 team created an instance profile and attached it to their EC2 instance during launch.
```

---
layout: post
title: "OpenSearch MCP server: Empowering the Amazon EC2 team's droplet analysis"
authors:
    - arjunkumargiri
    - cgunawat
    - rithinp 
    - ylwu
date: 2025-08-18
categories:
    - technical-posts
meta_keywords: "OpenSearch MCP server, Amazon Q, EC2 droplet analysis, multi-cluster connectivity, operational automation, Model Context Protocol"

> meta_description: "Learn how the Amazon EC2 team implemented OpenSearch MCP server with Amazon Q to automate droplet failure analysis across multiple OpenSearch clusters, reducing investigation time from hours to minutes."

Reviewer 4: Updated meta: 

_meta_keywords:_  OpenSearch MCP server, droplet analysis, Model Context Protocol, Amazon EC2, Amazon Q,  IAM authentication, droplet failure, performance optimization, AI-assisted operations, multi-cluster connectivity

_meta_description:_  Discover how Amazon EC2 team leveraged OpenSearch MCP server and Amazon Q to revolutionize droplet quality investigations, enabling efficient multi-cluster data analysis and reducing troubleshooting time from hours to minutes
